framework/core/js/src/common/components/ModalManager.tsx

```diff
@@ -10,90 +10,154 @@ interface IModalManagerAttrs {
-        {!!Tag && (
-          <Tag
-            key={modal?.key}
-            {...modal.attrs}
-            animateShow={this.animateShow.bind(this)}
-            animateHide={this.animateHide.bind(this)}
-            state={this.attrs.state}
-          />
-        )}
-      </div>
-    );
+    return this.attrs.state.modalList.map((modal: any) => {
+      const Tag = modal?.componentClass;
+
+      // Make stackable modals
+      const zIndex = 9999 + modal.key;
+
+      return (
+        <div className="ModalManager modal" modal-key={modal.key} style={{ zIndex }}>
+          {!!Tag && (
+            <Tag
+              key={modal?.key}
+              {...modal.attrs}
+              animateShow={this.animateShow.bind(this)}
+              animateHide={this.animateHide.bind(this)}
+              state={this.attrs.state}
+            />
+          )}
+
+          <div className="backdrop" key={`backdrop-${modal.key}`} onclick={this.handleBackdropClick.bind(this)} />
+        </div>
+      );
+    });
   }
 
   oncreate(vnode: Mithril.VnodeDOM<IModalManagerAttrs, this>): void {
     super.oncreate(vnode);
 
-    // Ensure the modal state is notified about a closed modal, even when the
-    // DOM-based Bootstrap JavaScript code triggered the closing of the modal,
-    // e.g. via ESC key or a click on the modal backdrop.
-    this.$().on('hidden.bs.modal', this.attrs.state.close.bind(this.attrs.state));
+    // Register keyup
+    document.body.addEventListener('keyup', this.handleEscPress.bind(this));
+  }
 
-    this.focusTrap = createFocusTrap(this.element as HTMLElement, { allowOutsideClick: true });
+  onbeforeremove(vnode: Mithril.VnodeDOM<IModalManagerAttrs, this>): void {
+    document.body.removeEventListener('keyup', this.handleEscPress);
   }
 
   onupdate(vnode: Mithril.VnodeDOM<IModalManagerAttrs, this>): void {
     super.onupdate(vnode);
 
     requestAnimationFrame(() => {
       try {
-        if (this.modalShown) this.focusTrap!.activate?.();
-        else this.focusTrap!.deactivate?.();
+        // Get current dialog key
+        const dialogKey = this?.attrs?.state?.modal?.key;
+
+        // Deactivate focus trap if there's a new dialog/closed
+        if (this.focusTrap && this.lastSetFocusTrap !== dialogKey) {
+          this.focusTrap!.deactivate?.();
+        }
+
+        // Activate focus trap if there's a new dialog which is not trapped yet
+        if (this.dialogElement && this.lastSetFocusTrap !== dialogKey) {
+          this.focusTrap = createFocusTrap(this.dialogElement as HTMLElement, { allowOutsideClick: true });
+
+          this.focusTrap!.activate?.();
+        }
+
+        // Update key of current opened modal
+        this.lastSetFocusTrap = dialogKey;
       } catch {
         // We can expect errors to occur here due to the nature of mithril rendering
       }
     });
   }
 
+  /**
+   * Get current active dialog
+   */
+  private get dialogElement(): HTMLElement {
```

Comment:
Could we rename this to more suitable match the doc block?


```suggestion
  private get activeDialogElement(): HTMLElement {
```